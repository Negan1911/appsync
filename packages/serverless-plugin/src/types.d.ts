declare module 'serverless-appsync-plugin' {
  import Serverless from 'serverless'
  import Plugin from 'serverless/classes/Plugin'
  class AppSync implements Plugin {
    hooks: Plugin['hooks']
    commands: Plugin['commands']
    serverless: Serverless
    validateSchemas: () => void
    constructor(serverless: Serverless, options: Serverless.Options)
  }

  export = AppSync
}

declare module 'schemaglue' {
  const options: { ignore: string }
  const res: { schema: string }
  const _default: (path: string, opt: typeof options) => typeof res
  export = _default
}
