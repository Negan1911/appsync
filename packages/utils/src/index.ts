import {
  parse,
  print,
  visit,
  Source,
  FieldDefinitionNode,
  DirectiveNode,
  StringValueNode,
  NameNode,
  IntValueNode,
  FloatValueNode,
  BooleanValueNode,
  EnumValueNode,
} from 'graphql'

export type PrimitiveValueNode =
  | IntValueNode
  | FloatValueNode
  | StringValueNode
  | BooleanValueNode
  | EnumValueNode

type SchemaDefinitions = {
  name: string
  field: string
  parent: string
  handler: string
}

interface NamableNode {
  name: NameNode
}

class Utils {
  /**
   * Returns an understandable error and a way to found on which node happened.
   * @internal
   * @param node Node where the error happened, will be formatted on the error.
   * @param message Description of the error.
   * @returns Formatted error.
   */
  static LocalizedError<Node extends NamableNode>(
    node: Node | Array<Node>,
    message: string
  ): Error {
    if (Array.isArray(node)) {
      const name = node.reduce((msg, _) => {
        if (!_?.name?.value) return msg
        return msg.length ? [msg, _.name.value].join(' > ') : _.name.value
      }, '')

      return new Error(name ? `[${name}]: ${message}` : message)
    }

    const name = (node as Node)?.name?.value
    return new Error(name ? `[${name}]: ${message}` : message)
  }

  /**
   * Returns an understandable error and a way to found on which node happened.
   * @internal
   * @param directive Directive node where the error happened.
   * @param field Field node where the error happened.
   * @param message description of the error.
   * @returns Formatted error.
   */
  static LocalizedFieldError(
    directive: DirectiveNode,
    field: FieldDefinitionNode,
    message: string
  ) {
    return Utils.LocalizedError([field, directive], message)
  }

  /**
   * Returns a directive argument value.
   * @internal
   * @template Type Type of the argument value.
   * @param directive Directive where the value belongs.
   * @param field Field where the directive belongs.
   * @param name Name of the argument to look for.
   * @param type Type of the argument value, validates against `Type`.
   * @param required If required, fails if not found.
   * @returns Argument type or null if not found (and not required).
   */
  static GetArgument<Type extends PrimitiveValueNode>(
    directive: DirectiveNode,
    field: FieldDefinitionNode,
    name: string,
    type: Type['kind'],
    required?: boolean
  ): Type['value'] | null {
    if (directive.arguments) {
      const argument = directive.arguments.find((_) => _.name.value === name)
      // Check if required
      if (!argument && required) {
        throw Utils.LocalizedFieldError(
          directive,
          field,
          `is missing required argument (${name}).`
        )
      }

      if (argument && argument.value) {
        const arg_type = argument.value.kind
        const arg_value = (argument.value as Type).value

        if (type && type !== arg_type) {
          const exp_t = type.replace('Value', '')
          const rcv_t = arg_type.replace('Value', '')

          throw Utils.LocalizedFieldError(
            directive,
            field,
            `Argument ${name} expected to be of type ${exp_t} but received ${rcv_t} instead.`
          )
        }

        return arg_value
      }

      return null
    }

    throw Utils.LocalizedFieldError(directive, field, 'Has no arguments.')
  }

  /**
   * Returns a specific directive from a field.
   * @internal
   * @param field Field where the directive belongs.
   * @param name Name of the directive to look for.
   * @returns Directive Node or undefined if not found
   */
  static GetDirective(field: FieldDefinitionNode, name: string) {
    if (field && field.directives) {
      return field.directives.find((d) => d.name.value === name)
    }

    return undefined
  }
}

export class GraphQLParser {
  /**
   * Extract definitions from a schema, and returns a useable schema.
   * @param schema Schema as string
   * @returns Tuple, first parameter a list of definitions taken from `@AWSLambda` directives, 2nd parameter is a clean schema
   */
  static ExtractFromSchema(
    schema: string | Source
  ): [Array<SchemaDefinitions>, string] {
    const deleted: Array<string> = []
    const definitions: Array<SchemaDefinitions> = []
    const ast = visit(parse(schema), {
      FieldDefinition: {
        enter(node, key, parent, path, ancestors) {
          const d = Utils.GetDirective(node, 'AWSLambda')
          if (d) {
            const field = node.name.value
            const parent = (ancestors[2] as FieldDefinitionNode).name.value
            const name = [parent, field].join('_')
            const handler = Utils.GetArgument<StringValueNode>(
              d,
              node,
              'handler',
              'StringValue',
              true
            )

            handler && definitions.push({ parent, name, handler, field })

            return {
              ...node,
              // Delete @AWSLambda from the directives list
              directives: node.directives?.filter(
                (_) => _.name.value !== 'AWSLambda'
              ),
            }
          }

          return null
        },
      },
    })

    return [
      definitions,
      print(
        visit(ast, {
          ObjectTypeDefinition: {
            enter(node) {
              if (!node?.fields?.length && deleted.includes(node.name.value)) {
                return null
              }

              return false
            },
          },
        })
      ),
    ]
  }
}
