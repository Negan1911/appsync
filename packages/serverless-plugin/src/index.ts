import fs from 'fs'
import path from 'path'
import glue from 'schemaglue'
import Serverless from 'serverless'
import { GraphQLParser } from '@appsync/utils'
import Service from 'serverless/classes/Service'
import AppSync = require('serverless-appsync-plugin')
const templates = path.join(
  require.resolve('@appsync/utils'),
  '..',
  '..',
  'templates'
)

type AppsyncConfig = { schemaDir?: string; schema?: string }

function ProcessSchema(servicePath: string, appsync: AppsyncConfig) {
  // Target of where the final schema (without special directive) will be.
  const schemapath = path.join(servicePath, '.serverless_schema.graphql')
  let raw_schema = null
  // Validates schemaDir input
  if (appsync.schemaDir) {
    // Base from where get the .graphql files.
    const basepath = path.join(servicePath, appsync.schemaDir)
    raw_schema = glue(basepath, { ignore: '**/!(*.graphql)' }).schema
  } else if (appsync.schema) {
    // Base from where get the .graphql files.
    raw_schema = fs.readFileSync(
      path.join(servicePath, appsync.schemaDir as string),
      'utf8'
    )
  } else
    throw new Error(
      'Invalid configuration: neither "schemaDir" nor "schema" config was found'
    )

  // Processed
  const [values, schema] = GraphQLParser.ExtractFromSchema(raw_schema)
  fs.writeFileSync(schemapath, schema, 'utf-8')

  return values
}

type AppSyncOptions = Serverless.Options & { serviceRole?: string }
type ServerlessInit = Serverless & { service: Service & { functions: object } }

export = class LambdaGraphQLPlugin extends AppSync {
  constructor(serverless: ServerlessInit, options: AppSyncOptions) {
    const serviceRole = options.serviceRole || 'AppSyncServiceRole'
    const functions = serverless.service.functions
    const entries = ProcessSchema(
      serverless.config.servicePath,
      serverless.service.custom.appSync
    )

    serverless.service.functions = {
      ...functions,
      ...entries.reduce((all, _) => {
        return { ...all, [_.name]: { name: _.name, handler: _.handler } }
      }, {}),
    }

    serverless.service.custom.appSync = {
      ...serverless.service.custom.appSync,
      schema: '.serverless_schema.graphql',
      dataSources: [
        ...(serverless.service.custom.appSync.dataSources || []),
        ...entries.map((_) => ({
          name: _.name,
          type: 'AWS_LAMBDA',
          config: {
            functionName: _.name,
            serviceRoleArn: { 'Fn::GetAtt': [serviceRole, 'Arn'] },
          },
        })),
      ],
      mappingTemplates: [
        ...(serverless.service.custom.appSync.mappingTemplates || []),
        ...entries.map((_) => ({
          dataSource: _.name,
          type: _.parent,
          field: _.field,
          request: path.join(templates, 'request.vtl'),
          response: path.join(templates, 'response.vtl'),
        })),
      ],
    }

    super(serverless, options)

    this.commands = {
      ...this.commands,
      validate: {
        usage: 'Validate your GraphQL Schema',
        lifecycleEvents: ['validate'],
      },
      'generate-types': {
        usage: 'Generate output GraphQL Schema',
        lifecycleEvents: ['gentype'],
      },
    }

    this.hooks = {
      ...this.hooks,
      'validate:validate': () => this.validateSchemas(),
      'gentype:gentype': () => {
        ProcessSchema(
          this.serverless.config.servicePath,
          this.serverless.service.custom.appSync
        )
      },
    }
  }
}
