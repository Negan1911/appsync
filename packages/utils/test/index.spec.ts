import { GraphQLParser } from '../src'

describe('ExtractFromSchema', () => {
  describe('Properly extract', () => {
    const schema = `
      type Author {
        name: String
      }

      type Todo {
        id: String!
        text: String!
        title: String!
        author: Author!
        @AWSLambda(handler: "Controllers/Todo/Author/index.default")
      }

      type Query {
        GetTodo(id: String!): Todo!
        @AWSLambda(handler: "Controllers/Todo/GetTodo.default")
        SearchTodo(search: String!): [Todo!]
        @AWSLambda(handler: "Controllers/Todo/SearchTodo.default")
      }

      schema {
        query: Query
      }
    `

    it('Should print a clean schema', () => {
      const [, clean_schema] = GraphQLParser.ExtractFromSchema(schema)
      expect(clean_schema).toMatchSnapshot()
    })

    it('Should properly fetch fields', () => {
      const [config] = GraphQLParser.ExtractFromSchema(schema)
      expect(config).toHaveLength(3)
      expect(config).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            parent: 'Todo',
            name: 'Todo_author',
            handler: 'Controllers/Todo/Author/index.default',
          }),
          expect.objectContaining({
            parent: 'Query',
            name: 'Query_GetTodo',
            handler: 'Controllers/Todo/GetTodo.default',
          }),
          expect.objectContaining({
            parent: 'Query',
            name: 'Query_SearchTodo',
            handler: 'Controllers/Todo/SearchTodo.default',
          }),
        ])
      )
    })
  })

  describe('Required Argument not found', () => {
    const schema = `
      type Author {
        name: String
      }

      type Todo {
        id: String!
        text: String!
        title: String!
        author: Author!
        @AWSLambda(handlr: "Controllers/Todo/Author/index.default")
      }

      type Query {
        GetTodo(id: String!): Todo!
        @AWSLambda(handler: "Controllers/Todo/GetTodo.default")
        SearchTodo(search: String!): [Todo!]
        @AWSLambda(handler: "Controllers/Todo/SearchTodo.default")
      }

      schema {
        query: Query
      }
    `

    it('Should fail with a proper "missing handler" error.', () => {
      expect(() => GraphQLParser.ExtractFromSchema(schema)).toThrowError(
        '[author > AWSLambda]: is missing required argument (handler).'
      )
    })
  })

  describe('Incorrect Argument type', () => {
    const schema = `
      type Author {
        name: String
      }

      type Todo {
        id: String!
        text: String!
        title: String!
        author: Author!
        @AWSLambda(handler: "Controllers/Todo/Author/index.default")
      }

      type Query {
        GetTodo(id: String!): Todo!
        @AWSLambda(handler: 3)
        SearchTodo(search: String!): [Todo!]
        @AWSLambda(handler: "Controllers/Todo/SearchTodo.default")
      }

      schema {
        query: Query
      }
    `

    it('Should fail with a proper "incorrect type" error.', () => {
      expect(() => GraphQLParser.ExtractFromSchema(schema)).toThrowError(
        '[GetTodo > AWSLambda]: Argument handler expected to be of type String but received Int instead.'
      )
    })
  })
})
