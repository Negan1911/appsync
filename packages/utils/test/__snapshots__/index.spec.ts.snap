// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtractFromSchema Properly extract Should print a clean schema 1`] = `
"type Author

type Todo {
  author: Author!
}

type Query {
  GetTodo(id: String!): Todo!
  SearchTodo(search: String!): [Todo!]
}

schema {
  query: Query
}
"
`;
